### PR TITLE
[pigeon] Remove heap allocation in generated C++ code

### PR DESCRIPTION
**Pull request by loic-sharma**
Please let me know if I goofed anything, this is my first time contributing to Pigeon!

Addresses https://github.com/flutter/flutter/issues/144042

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/pigeon/example/app/windows/runner/messages.g.cpp [hunk 1]
```diff
@@ -1,7 +1,7 @@
 // Copyright 2013 The Flutter Authors. All rights reserved.
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
-// Autogenerated from Pigeon, do not edit directly.
+// Autogenerated from Pigeon (v17.0.0), do not edit directly.
```
Comment (loic-sharma) [inline packages/pigeon/example/app/windows/runner/messages.g.cpp:4]: FYI, Pigeon also updated other platforms' comment to add a Pigeon version. I reverted the other platforms' changes.

File: packages/pigeon/platform_tests/test_plugin/example/windows/flutter/CMakeLists.txt [hunk 2]
```diff
@@ -10,6 +10,11 @@ include(${EPHEMERAL_DIR}/generated_config.cmake)
 # https://github.com/flutter/flutter/issues/57146.
 set(WRAPPER_ROOT "${EPHEMERAL_DIR}/cpp_client_wrapper")
 
+# Set fallback configurations for older versions of the flutter tool.
+if (NOT DEFINED FLUTTER_TARGET_PLATFORM)
+  set(FLUTTER_TARGET_PLATFORM "windows-x64")
+endif()
```
Comment (loic-sharma) [inline packages/pigeon/platform_tests/test_plugin/example/windows/flutter/CMakeLists.txt:16]: The Flutter tool automatically does this migration to add Windows Arm64 support.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -17,6 +17,8 @@ Additional resources specific to the packages repository:
   As explained in the Flutter guide,
   [**PRs need tests**](https://github.com/flutter/flutter/wiki/Tree-hygiene#tests), so
   this is critical to read before submitting a plugin PR.
+- [Plugin tool](https://github.com/flutter/packages/tree/main/script/tool),
```
Comment (stuartmorgan) [inline CONTRIBUTING.md:20]: This is covered in other links above; is there a specific reason to add it? I deliberately try to keep this list short to focus attention on the most important entry points.

File: packages/pigeon/lib/cpp_generator.dart [hunk 4]
```diff
@@ -871,11 +871,9 @@ class CppSourceGenerator extends StructuredGenerator<CppOptions> {
           scope: api.name,
           returnType: _voidType,
           parameters: parameters, body: () {
-        const String channel = 'channel';
         indent.writeln(
             'const std::string channel_name = "${makeChannelName(api, func, dartPackageName)}";');
-        indent.writeln(
-            'auto channel = std::make_unique<BasicMessageChannel<>>(binary_messenger_, '
+        indent.writeln('BasicMessageChannel<> channel(binary_messenger_, '
```
Comment (stuartmorgan) [inline packages/pigeon/lib/cpp_generator.dart:876]: These changes should ideally have a Dart unit test; that's generally how we handle changes that affect generation in some desirable way, but don't change observed behavior.

File: packages/pigeon/example/app/windows/runner/messages.g.cpp (hunk 1 above)
Comment (stuartmorgan) [reply packages/pigeon/example/app/windows/runner/messages.g.cpp:4]: (Probably we need to do a pass over this package's CONTRIBUTING.md with updated info.)

File: CONTRIBUTING.md (hunk 3 above)
Comment (loic-sharma) [reply CONTRIBUTING.md:20]: The current docs feel a bit fragmented. From skimming around I missed this tool existed. I'll revert this though.

File: packages/pigeon/example/app/windows/runner/messages.g.cpp (hunk 1 above)
Comment (loic-sharma) [reply packages/pigeon/example/app/windows/runner/messages.g.cpp:4]: Ah, should I update this document then?

https://github.com/flutter/packages/blob/fe6111ba54651ea793879070f90b8df9a10feb58/packages/pigeon/example/app/README.md?plain=1#L6-L9

FYI, `tool/generate.dart` and does not appear to work on my machine. I'll debug

https://github.com/flutter/packages/tree/main/packages/pigeon/example/app#pigeon_example_app

Hm, I did try `tool/generate.dart` but that hangs on my machine. I'll debug it...

Comment (loic-sharma) [reply packages/pigeon/example/app/windows/runner/messages.g.cpp:4]: I have a draft to fix the hang here: https://github.com/flutter/packages/pull/6198

File: CONTRIBUTING.md (hunk 3 above)
Comment (loic-sharma) [reply CONTRIBUTING.md:20]: Reverted!

File: packages/pigeon/lib/cpp_generator.dart (hunk 4 above)
Comment (loic-sharma) [reply packages/pigeon/lib/cpp_generator.dart:876]: Gotcha, I added this but let me know if this isn't what you had in mind: https://github.com/flutter/packages/pull/6196/files#diff-e8ad868911157f965b470ee46bff77cea5651550151081a6114d6a445ac2cc2fR2118

**Comment by loic-sharma** [comment]
@stuartmorgan @tarrinneal Gentle reminder for review :)

**Review by tarrinneal — APPROVED** [review]
If you could sync the latest changes, this all seems good to me.

**Pull request merged**